src/coreclr/tools/Common/JitInterface/CorInfoImpl.cs

```diff
@@ -1025,19 +1055,92 @@ private MethodSignatureFlags PlatformDefaultUnmanagedCallingConvention()
         {
-            MethodSignatureFlags unmanagedCallConv = HandleToObject(method).GetPInvokeMethodMetadata().Flags.UnmanagedCallingConvention;
+            suppressGCTransition = false;
+            MethodSignatureFlags callConv = methodDesc.Signature.Flags & MethodSignatureFlags.UnmanagedCallingConventionMask;
+            if (callConv == MethodSignatureFlags.None)
+            {
+                if (methodDesc.IsPInvoke)
+                {
+                    suppressGCTransition = methodDesc.IsSuppressGCTransition();
+                    MethodSignatureFlags unmanagedCallConv = methodDesc.GetPInvokeMethodMetadata().Flags.UnmanagedCallingConvention;
 
-            if (unmanagedCallConv == MethodSignatureFlags.None)
-                unmanagedCallConv = PlatformDefaultUnmanagedCallingConvention();
+                    if (unmanagedCallConv == MethodSignatureFlags.None)
+                        unmanagedCallConv = PlatformDefaultUnmanagedCallingConvention();
 
-            // Verify that it is safe to convert MethodSignatureFlags.UnmanagedCallingConvention to CorInfoUnmanagedCallConv via a simple cast
-            Debug.Assert((int)CorInfoUnmanagedCallConv.CORINFO_UNMANAGED_CALLCONV_C == (int)MethodSignatureFlags.UnmanagedCallingConventionCdecl);
-            Debug.Assert((int)CorInfoUnmanagedCallConv.CORINFO_UNMANAGED_CALLCONV_STDCALL == (int)MethodSignatureFlags.UnmanagedCallingConventionStdCall);
-            Debug.Assert((int)CorInfoUnmanagedCallConv.CORINFO_UNMANAGED_CALLCONV_THISCALL == (int)MethodSignatureFlags.UnmanagedCallingConventionThisCall);
+                    // Verify that it is safe to convert MethodSignatureFlags.UnmanagedCallingConvention to CorInfoCallConvExtension via a simple cast
+                    Debug.Assert((int)CorInfoCallConvExtension.C == (int)MethodSignatureFlags.UnmanagedCallingConventionCdecl);
+                    Debug.Assert((int)CorInfoCallConvExtension.Stdcall == (int)MethodSignatureFlags.UnmanagedCallingConventionStdCall);
+                    Debug.Assert((int)CorInfoCallConvExtension.Thiscall == (int)MethodSignatureFlags.UnmanagedCallingConventionThisCall);
 
-            return (CorInfoUnmanagedCallConv)unmanagedCallConv;
+                    return (CorInfoCallConvExtension)unmanagedCallConv;
+                }
+                else
+                {
+                    Debug.Assert(methodDesc.IsUnmanagedCallersOnly);
+                    CustomAttributeValue<TypeDesc> unmanagedCallersOnlyAttribute = ((EcmaMethod)methodDesc).GetDecodedCustomAttribute("System.Runtime.InteropServices", "UnmanagedCallersOnlyAttribute").Value;
+                    return GetUnmanagedCallingConventionFromAttribute(unmanagedCallersOnlyAttribute);
+                }
+            }
+            return GetUnmanagedCallConv(methodDesc.Signature, out suppressGCTransition);
+        }
+
+        private CorInfoCallConvExtension GetUnmanagedCallConv(MethodSignature signature, out bool suppressGCTransition)
```

Comment:
This is part of my question for the optional out. It doesn't seem like we optimize this path if null is provided so making callers be explicit about asking for this detail seems appropriate.